jet_tracking/sim.py

```diff
@@ -0,0 +1,195 @@
+from types import SimpleNamespace
+
+import numpy as np
+from ophyd import Signal
+from ophyd.sim import SynAxis, SynSignal
+
+
+class BeamSignal(Signal):
+    """
+    Class to represent the energy of the beam
+
+    The setpoint can be entered by the user simply by use ``.put``. After that,
+    two kinds of noise will be added. First, a gaussian distribution of with
+    ``sigma`` equal to the :param:`.noise` is added. Secondly, in order to
+    simulate frequent drops in the beam, certain ``.get`` attempts will report
+    a zero value as if the shot was dropped.
+
+    Parameters
+    ----------
+    noise: float, optional
+        Width of distribution of values around the energy setpoint. For
+        instance, if set to 1.0, 68% of values will lie within 1.0 of the
+        setpoint.
+
+    drop_rate: float, optional
+        Probability that a shot will be dropped and report 0. as the energy.
+        This should be a value between 0 and 1, with 1 being a 100% chance that
+        the shot is counted as dropped.
+
+    args, kwargs:
+        Passed to ``ophyd.Signal`` constructor
+    """
+    def __init__(self, *args, noise=0., drop_rate=0., **kwargs):
+        super().__init__(*args, **kwargs)
+        self.noise = noise
+        self.drop_rate = drop_rate
+
+    def get(self):
+        """Get the signal. Add noise or drop based on settings"""
+        if np.random.uniform(0, 1) > self.drop_rate:
+            value = super().get()
+            if self.noise:
+                value += np.random.normal(0., self.noise)
+            return value
+        return 0.
+
+
+def generate_simulation(motor_precision=3,
+                        beam_energy=2.5,
+                        energy_noise=0.3,
+                        beam_position=0.,
+                        drop_rate=0.15,
+                        adu_max=15000,
+                        adu_sigma=1500,
+                        adu_floor=200,
+                        jet_width=0.01,
+                        jet_noise=0.005,
+                        ):
+    """
+    Generate a simulation of the injector, gas detector and CSPAD intensity
+
+    Parameters
+    ----------
+    motor_precision: int, optional
+        Limit the number of significant digits of the injector position. Stored
+        on ``inj_x``.
+
+    beam_energy: float, optional
+        The starting energy of the beam in the simulation. Modify by setting
+        ``beam``
+
+    drop_rate: float, optional
+        The probability that the energy of a shot will be zero. Stored on the
+        ``beam`` object.
+
+    energy_noise: float, optional
+        The value of a single standard deviation for flucuations in beam
+        energy. Stored on the ``beam`` object.
+
+    beam_position: float, optional
+        Position of the beam. This is where the center of the Gaussian
+        distribution will be for the ``cspad_adu``.
+
+    adu_max: float, optional
```

Comment:
Could we just keep ADU normalized 0-1, since we're ignoring the gas detector normalization aspect? Real ADU numbers are on the order of 10^9 or so IIRC